src/lib/rnp.c

```diff
@@ -1147,6 +1149,148 @@ rnp_decrypt_file(rnp_ctx_t *ctx, const char *f, const char *out)
     return ret ? RNP_SUCCESS : RNP_ERROR_GENERIC;
 }
 
+typedef struct pgp_parse_handler_param_t {
+    char in[PATH_MAX];
+    char out[PATH_MAX];
+} pgp_parse_handler_param_t;
+
+/* process the pgp stream */
+
+bool
+rnp_parse_handler_dest(pgp_parse_handler_t *handler, pgp_dest_t *dst, const char *filename)
+{
+    pgp_parse_handler_param_t *param = handler->param;
+
+    if (!param) {
+        return false;
+    }
+
+    if (strlen(param->out) > 0) {
+        return init_file_dest(dst, param->out) == RNP_SUCCESS;
+    } else {
+        return init_stdout_dest(dst) == RNP_SUCCESS;
+    }
+}
+
+int
+rnp_process_stream(rnp_ctx_t *ctx, const char *in, const char *out)
+{
+    pgp_source_t               src;
+    rnp_result_t               err;
+    pgp_parse_handler_t *      handler = NULL;
+    pgp_parse_handler_param_t *param = NULL;
+    int                        result;
+
+    if (in && (strlen(in) > sizeof(param->in))) {
+        (void) fprintf(stderr, "rnp_process_stream: too long input path\n");
+        return RNP_FAIL;
+    }
+
+    if (out && (strlen(out) > sizeof(param->out))) {
+        (void) fprintf(stderr, "rnp_process_stream: too long output path\n");
+        return RNP_FAIL;
+    }
+
+    err = in ? init_file_src(&src, in) : init_stdin_src(&src);
+
+    if (err != RNP_SUCCESS) {
+        return RNP_FAIL;
+    }
+
+    if ((handler = calloc(1, sizeof(*handler))) == NULL) {
+        result = RNP_FAIL;
+        goto finish;
+    }
+
+    if ((param = calloc(1, sizeof(*param))) == NULL) {
+        result = RNP_FAIL;
+        goto finish;
+    }
+
+    if (in) {
+        strcpy(param->in, in);
+    } else {
+        param->in[0] = 0;
+    }
+
+    if (out) {
+        strcpy(param->out, out);
+    } else {
+        param->out[0] = 0;
+    }
+
+    handler->passphrase_provider = &ctx->rnp->passphrase_provider;
+    handler->dest_provider = rnp_parse_handler_dest;
+    handler->param = param;
+
+    err = process_pgp_source(handler, &src);
+    if (err == RNP_SUCCESS) {
+        result = RNP_OK;
+    } else {
+        (void) fprintf(stderr, "processing failed: error 0x%x\n", err);
+        result = RNP_FAIL;
+    }
+
+finish:
+    src_close(&src);
+    free(handler);
+    free(param);
```

Comment:
you can `free` NULL pointer here

Reply:
Nothing wrong with calling free on a NULL pointer.

Reply:
Yeah, free(NULL) works pretty well.

Reply:
yes, it is safe but looks strange for me. But I agree that it is a style question

Reply:
A particular implementation isn't really relevant, the C standard itself says `if ptr is a null pointer, no action occurs` to cover causes just like this.